kart/cli.py

```diff
@@ -233,6 +246,210 @@ def push(ctx, do_progress, args):
+                
+                sys.stdin = os.fdopen(fds[0], "r")
+                sys.stdout = os.fdopen(fds[1], "w")
+                sys.stderr = os.fdopen(fds[2], "w")
+
+                try:
+                    calling_environment = json.loads(payload)
+                except (TypeError, ValueError) as e:
+                    click.echo(
+                        "kart helper: Unable to read command from kart_cli_helper", e
+                    )
+                else:
+                    sys.argv[1:] = calling_environment["argv"][1:]
+                    kart_helper_log.write(f"PID={calling_environment['pid']} CWD={os.getcwd()} CMD={' '.join(calling_environment['argv'])}\n")
+                    os.environ.clear()
+                    os.environ.update(
+                        {**calling_environment["environ"], **required_environment, 'KART_HELPER_PID': str(os.getppid())}
+                    )
+
+                    import ctypes
+                    import ctypes.util
+
+                    libc = ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True)
+                    # 
+                    if is_darwin:
+                        SETVAL = 8
+                    elif is_linux:
+                        SETVAL = 16
+
+                    class struct_semid_ds(ctypes.Structure):
+                        pass
+
+                    class struct_semun(ctypes.Union):
+                        _fields_ = [
+                            ("val", ctypes.c_uint32),
+                            (
+                                "buf",
+                                ctypes.POINTER(struct_semid_ds),
+                            ),
+                            (
+                                "array",
+                                ctypes.POINTER(ctypes.c_uint16),
+                            ),
+                        ]
+
+                    libc.semctl.argtypes = [
+                        ctypes.c_int,
+                        ctypes.c_int,
+                        ctypes.c_int,
+                        struct_semun,
+                    ]
+                    libc.semget.argtypes = [ctypes.c_int, ctypes.c_int, ctypes.c_int]
+                    semid = calling_environment["semid"]
+                    try:
+                        cli()
+                    except SystemExit as system_exit:
+                        """exit is called in the commands but we ignore as we need to clean up the caller"""
+
+                        # TODO - do we ever see negative exit codes from cli (git etc)?
+                        libc.semctl(
+                            semid, 0, SETVAL, struct_semun(val=system_exit.code + 1000)
+                        )
+
+                    except Exception as e:
+                        traceback.print_exc(file=sys.stdout)
+                        print(
+                            f"kart helper: unhandled exception [{e}]"
+                        )  # TODO - should ext-run capture/handle this?
```

Comment:
Do you need both of these ways of printing the exception?
Should it be to stdout or to stderr?

Reply:
Will drop the second here. This is more something that I was wanting some feedback on with the ext-run command, whether it should be doing something with the command and capture an error as a different result code. I'll create an issue around that.